### PR TITLE
Add some config options to the UART driver

### PR DESCRIPTION
**Pull request by bjoernQ**
This implement configuring the UART / serial driver and adds examples to show how to do that.

Additionally reading never worked on ESP32-S2 - this is fixed now

### CODE REVIEW

File: esp32-hal/src/lib.rs [hunk 1]
```diff
@@ -34,6 +34,11 @@ pub mod analog {
     pub use esp_hal_common::analog::{AvailableAnalog, SensExt};
 }
 
+/// Serial Configuration
+pub mod serial {
+    pub use esp_hal_common::serial::*;
+}
```
Comment (jessebraham) [inline esp32-hal/src/lib.rs:40]: Is there a reason we can't just do `pub use esp_hal_common::serial;` here, like we do with the other driver modules?

(Same for the 3 remaining chips)

File: esp-hal-common/src/serial.rs [hunk 2]
```diff
@@ -26,7 +158,35 @@ impl<T> Serial<T>
...
+        TX: OutputPin<OutputSignal = OutputSignal>,
+        RX: InputPin<InputSignal = InputSignal>,
+        CTS: InputPin<InputSignal = InputSignal>,
+        RTS: OutputPin<OutputSignal = OutputSignal>,
+    >(
+        uart: T,
+        config: Option<Config>,
+        pins: Option<Pins<TX, RX, CTS, RTS>>,
+        clocks: &Clocks,
+    ) -> Result<Self, Error> {
```
Comment (jessebraham) [inline esp-hal-common/src/serial.rs:172]: Can this actually fail? Is the `Result<_>` return type necessary?

File: esp-hal-common/src/serial.rs [hunk 3]
```diff
@@ -167,20 +442,102 @@ pub trait Instance {
...
+    ) {
+        if pins.is_some() {
+            let pins = pins.unwrap();
+
+            if pins.tx.is_some() {
+                pins.tx
+                    .unwrap()
+                    .set_to_push_pull_output()
+                    .connect_peripheral_to_output(self.tx_signal());
+            }
```
Comment (jessebraham) [inline esp-hal-common/src/serial.rs:463]: Using `if let Some(tx) = pins.tx {}` would eliminate the `.unwrap()` here. Not a huge deal but probably a bit more idiomatic.

**Review by jessebraham — COMMENTED** [review]
Left a few review comments, nothing too major. If you don't want to address them now (or ever 😁), that's fine.

As discussed offline it'd be nice to eliminate the `NoPin` type if possible, that that can be done in subsequent PR(s) if we do move forward with that.

File: esp32-hal/src/lib.rs (hunk 1 above)
Comment (bjoernQ) [reply esp32-hal/src/lib.rs:40]: I think for the docs ... but they look okay even with just the usual `pub use`

File: esp-hal-common/src/serial.rs (hunk 2 above)
Comment (bjoernQ) [reply esp-hal-common/src/serial.rs]: No - just copied the signature from the function that used to be here 👍

**Review by jessebraham — APPROVED** [review]
Thank you for addressing my comments, as well as making the changes we discussed yesterday. I think this is looking pretty good!

**Pull request merged**